internal/v4/api.go

```diff
@@ -1033,3 +1025,27 @@ func (h *ReqHandler) resolveId(f resolvedIdHandler) router.IdHandler {
 		return f(rid, isFullySpecified(id), w, req)
 	}
 }
+
+// auditSetPerms returns an audit entry to record later that the permissions of the given
+// entity have been set to the given ACLs.
+func (h *ReqHandler) auditSetPerms(id *router.ResolvedURL, read, write []string) *audit.Entry {
+	return h.auditSetUser(&audit.Entry{
+		Op:     audit.OpSetPerm,
+		Entity: &id.URL,
+		ACL: &audit.ACL{
+			Read:  read,
+			Write: write,
+		},
+	})
+}
+
+func (h *ReqHandler) auditSetUser(e *audit.Entry) *audit.Entry {
```

Comment:
The name of this is of the same form as auditSetPerms, but they do very different things - one returns a new audit entry that records a "set perms" operation; the other sets the user in an audit entry.

How about instead defining a method on ReqHandler to add an audit entry?

    func (h *ReqHandler) addAudit(e *audit.Entry) {
        if h.auth.Username == "" && !h.auth.Admin {
            panic("No auth set in ReqHandler")
        }
        e.User = h.auth.Username
        if h.auth.Admin && e.User == "" {
        	e.User = "admin"
        }
        h.Store.AddAudit(e)
        if testAddAuditCallback != nil {
            testAddAuditCallback(*e)
        }
    }

Then everywhere that adds an audit entry would use that method
and automatically gain the benefit of the user being added.